rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_equipment_associated_with_each_swh_distribution_system.py

```diff
@@ -0,0 +1,85 @@
+    """
+    This function gets all the SWH equipment connected to a SWH distribution system.  The information is stored in a dictionary where the keys are the SWH Distribution System Ids and values are a dictionary giving the ServiceWaterHeatingEquipment, and Pumps connected to the particular SWH distribution system.
+
+    Parameters
+    ----------
+    rmd: dict, RMD at RuleSetModelDescription level
+
+    Returns
+    -------
+    swh_and_equip_dict: A dictionary containing where the keys are the SWH Distribution System IDs and values are dictionaries where keys are the type of SWH equipment and values are the ids of the connected equipment.
+                        Example:  {"swh_distribution1":{"swh_heating_eq":["swh_eq1","swh_eq2"], "pumps":["p1"], "tanks":["t1"], "piping":["piping1"], "solar_thermal":[], "uses":["sp1_use","sp2_use"], "spaces_served":[sp1,sp2]}}
+
+    """
+
+    swh_and_equip_dict = {}
+    for distribution in find_all(
+        "$.service_water_heating_distribution_systems[*]", rmd
+    ):
+        swh_and_equip_dict[distribution["id"]] = SWHDistributionAssociations()
+
+        for tank in distribution.get("tanks", []):
+            swh_and_equip_dict[distribution["id"]].tanks.append(tank["id"])
+
+        for piping in distribution.get("service_water_piping", []):
+            queue = [piping]
+            piping_ids = []
+            # BFS approach
+            while queue:
+                current_piping = queue.pop(0)
+                piping_ids.append(current_piping["id"])
+                if current_piping.get("child"):
+                    queue.extend(current_piping["child"])
+            swh_and_equip_dict[distribution["id"]].piping.extend(piping_ids)
+
+        for pump in find_all("$.pumps[*]", rmd):
+            if (
+                pump.get("loop_or_piping")
+                in swh_and_equip_dict[distribution["id"]].piping
+            ):
+                swh_and_equip_dict[distribution["id"]].pumps.append(pump["id"])
+
+        for swh_equip in find_all("$.service_water_heating_equipment[*]", rmd):
+            if swh_equip.get("distribution_system") == distribution["id"]:
+                swh_and_equip_dict[distribution["id"]].swh_heating_eq.append(
+                    swh_equip["id"]
+                )
+                for solar_t in swh_equip.get("solar_thermal_systems", []):
+                    swh_and_equip_dict[distribution["id"]].solar_thermal.append(
+                        solar_t["id"]
+                    )
+
+    for swh_use in find_all(
+        "$.buildings[*].building_segments[*].zones[*].spaces[*].service_water_heating_uses[*]",
+        rmd,
+    ):
+        distribution_id = swh_use.get("served_by_distribution_system")
+        if distribution_id:
+            swh_and_equip_dict[distribution_id].uses.append(swh_use["id"])
+            swh_and_equip_dict[distribution_id].spaces_served.extend(
+                get_spaces_served_by_swh_use(rmd, swh_use["id"])
+            )
```

Comment:
Should add another loop to go over the `swh_and_equip_dict.spaces_served` to make sure unique names (or no duplications).